### DISCUSSION TITLE
Ustawianie sekretow w KV - ARM czy skrypty?

### DISCUSSION THREAD

**Discussion by filipendo125** [AzureCommunityPL/group#129] [Pomoc] [answered]
Przegladajac sobie rozne ARM-y ostatnio napotkalem na dokumentacje ustawiajaca w ARM-ie sekrety w KV. Wszystko jasne, nic skomplikowanego,ale zastanawiam sie, czy ustawianie kluczy w ARM-ie jest lepsze od ustawiania ich w jakis inny sposob - przykladowo PS, ktory iteruje sobie po parametrach i wpisuje je do KV.

Rozumiem, ze wpisanie tego w templatce gwarantuje Ci to, ze zaraz ktos nie przyjdzie i Ci tego nie zmieni (jednak ciezej jest zmienic tak duzy plik), ale z drugiej strony przy wpisywaniu przykladowo 10 sekretow, to juz bym sie zastanowil, czy ten template jest czytelny.

Jakie jest Wasze podejscie lub ktore podejscie rekomendujecie?

**Comment by wi5nia** [comment]
Ja jestem za [rekomendacja](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/key-vault-parameter?tabs=azure-cli) MSu w tej sprawie, czyli

1. CLI/PS - tworzyc KV i ustawiasz sekrety
2. Referencje przez plik parametrow dla ARMa

**Reply by filipendo125** [reply]
@wi5nia To rozwiazanie mowi bardziej o wykorzystywaniu tych rzeczy podczas deploymentu, ale zalozmy, ze tutaj po prostu chcesz wpisac rzeczy do KV niekoniecznie korzystajach z nich podczas deploymentu.

**Reply by kamil-mrzyglod** [reply]
@filipendo125 - nic Ciebie nie zmusi aby ustawiać sekrety via ARM Template. Równie dobrze mógłbyś mieć proces, że ktoś sekret ustawia z palca. To co pewnie warto rozważyć przy wyborze to podsumowanie @mgrabarz poniżej.

**Comment by mgrabarz** [comment, marked as answer]
Znów dorzucę od siebie ;)

Tworzenie kv i secretów przy pomocy ARMa jest wykonywane przez Azure control plane.
Robienie tego przez cli jest operacja data plane.

Ważne zalety pierwszego:
- Operacje control plane mogą omijać firewall i private linki dla kv odgrodzonych od świata.
- Nie wymagają kv policy bo działają na contributorze.
- Są audytowalne w historii deploymentów.

W przypadku drugiego:
- Audyt można dodać dodatkowo w kv
- Można zarobić bardziej granularne dostępu przez policy
- Trzeba tymczasowo whitelistowac

**Reply by kamil-mrzyglod** [reply]
`Nie wymagają kv policy bo działają na contributorze.` - jeśli mnie pamięc nie myli, pod warunkiem że zezwolisz na KV dostęp za pośrednictwem ARM ;)

**Reply by mgrabarz** [reply]
To nie kwestia rbac tylko sieciowa, dlatego napisałem "mogą omijać".

**Reply by kamil-mrzyglod** [reply]
True, nie pod tego bulleta się podpiąłem - RBAC nie ma tutaj nic do gadania 😄

**Comment by piotrgwiazda** [comment]
To zależy skąd bierzesz secrety? Jeśli to jakieś stałe, to ten ARM z
secretami leży w Gicie (?!)

W Terraform robimy tak, że Terraform stawia usługi i różne secrety do
usług, które uruchomił, wkłada do KV. Są to wygenerowane secrety więc nie
ma ich w Git. Są w statefile ale to inny temat.

Nie wiem do końca jakie są opcje dynamicznego generowania haseł itp w ARM
ale trzymałbym się zasady "Żadnych secretów w Gicie, do żadnych środowisk,
pod żadnym pozorem, bez wyjątków".



pt., 30 paź 2020, 17:40 użytkownik Kamil Mrzygłód <notifications@github.com>
napisał:

**Reply by kaluzaaa** [reply]
Co do poświadczeń i git - można jeśli szyfrujesz. Masz różne dobre rozwiązania ;-)
